### PR TITLE
Custom 404 and 500 error pages - Proposal

### PR DESCRIPTION
**Pull request by perwendel**

### CODE REVIEW

File: src/main/java/spark/CustomErrorPages.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public static Object getFor(int status, Request request, Response response) {
+
+        Object customRenderer = CustomErrorPages.getInstance().customPages.get(status);
+        Object customPage;
+
+        if (customRenderer instanceof String) {
+            customPage = customRenderer;
+        } else {
```
Comment (tipsy) [inline src/main/java/spark/CustomErrorPages.java]: Change to `if(customRenderer instanceof Route)` ?

File: src/main/java/spark/CustomErrorPages.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            try {
+                customPage = ((Route) customRenderer).handle(request, response);
+            } catch (Exception e) {
+                customPage = status == 400 ? NOT_FOUND : INTERNAL_ERROR;
+            }
+        }
+
```
Comment (tipsy) [inline src/main/java/spark/CustomErrorPages.java]: Remove line?

File: src/main/java/spark/RequestResponseFactory.java [hunk 3]
```diff
@@ -26,6 +26,13 @@
     private RequestResponseFactory() {
     }
 
+    /**
+     * Solely used for custom error pages.
```
Comment (tipsy) [inline src/main/java/spark/RequestResponseFactory.java]: I'm not sure this comment is helpful. Could the method be used for anything else? If not, call it something that contains the words "custom error pages", and remove the comment?

File: src/main/java/spark/Service.java [hunk 4]
```diff
@@ -323,6 +323,39 @@ public synchronized Service webSocketIdleTimeoutMillis(int timeoutMillis) {
...
+    }
+
+    /**
+     * Maps 500 internal server errors to the provided route.
+     */
+    public void internalServerError(Route route) {
+        CustomErrorPages.add(500, route);
+    }
+
+
```
Comment (tipsy) [inline src/main/java/spark/Service.java]: 2x empty  line

File: src/main/java/spark/RequestResponseFactory.java (hunk 3 above)
Comment (perwendel) [inline src/main/java/spark/RequestResponseFactory.java]: It's for specifying that it only should be used from there. But when thinking about it this class should no nothing on which classes that uses it. I'll re-write it

File: src/main/java/spark/CustomErrorPages.java (hunk 1 above)
Comment (perwendel) [inline src/main/java/spark/CustomErrorPages.java]: done

File: src/main/java/spark/CustomErrorPages.java (hunk 2 above)
Comment (perwendel) [inline src/main/java/spark/CustomErrorPages.java]: done

File: src/main/java/spark/RequestResponseFactory.java (hunk 3 above)
Comment (perwendel) [inline src/main/java/spark/RequestResponseFactory.java]: fixed

File: src/main/java/spark/Service.java (hunk 4 above)
Comment (perwendel) [inline src/main/java/spark/Service.java]: fixed

**Pull request merged**